File: breakpilot-idea/src/main/kotlin/debugger/CommandExecutor.kt
package debugger

import bridge.BridgeClient
import bridge.BridgeMessage
import bridge.MessageTypes
import com.intellij.execution.ProgramRunnerUtil
import com.intellij.execution.RunManager
import com.intellij.execution.RunnerAndConfigurationSettings
import com.intellij.execution.actions.ConfigurationContext
import com.intellij.execution.executors.DefaultDebugExecutor
import com.intellij.execution.runners.ExecutionEnvironmentBuilder
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiManager
import com.intellij.xdebugger.XDebuggerUtil

private data class SourceDebugTarget(
    val settings: RunnerAndConfigurationSettings,
    val sourceElement: String
)

class CommandExecutor(
    private val project: Project,
    private val bridge: BridgeClient,
    private val tracker: IdeSessionTracker,
    private val variableReader: VariableReader
) {
    fun handle(message: BridgeMessage) {
        when (message.type) {
            MessageTypes.AgentStartDebug -> ApplicationManager.getApplication().invokeLater { startDebug(message) }
            MessageTypes.AgentRequestStack -> requestStack(message)
            MessageTypes.AgentContinue -> execute(message, "continue") { it.resume() }
            MessageTypes.AgentPause -> execute(message, "pause") { it.pause() }
            MessageTypes.AgentStepOver -> execute(message, "step_over") { it.stepOver(false) }
            MessageTypes.AgentStepInto -> execute(message, "step_into") { it.stepInto() }
            MessageTypes.AgentStepOut -> execute(message, "step_out") { it.stepOut() }
            MessageTypes.AgentRunToLine -> runToLine(message)
            MessageTypes.AgentSetVariable -> setVariable(message)
            MessageTypes.AgentStopDebug -> execute(message, "stop_debug") { it.stop() }
            MessageTypes.AgentEvaluate -> evaluate(message)
            MessageTypes.AgentListRunConfigurations -> listRunConfigurations(message)
        }
    }

    private fun startDebug(message: BridgeMessage) {
        val runConfigName = message.runConfigName
        if (!runConfigName.isNullOrBlank()) {
            val settings = RunManager.getInstance(project).allSettings
                .firstOrNull { it.name == runConfigName }
            if (settings == null) {
                sendError(message, "start_debug", "RUN_CONFIG_NOT_FOUND", "Run configuration was not found: $runConfigName")
                return
            }
            try {
                executeWithOrigin(settings, message)
            } catch (error: Throwable) {
                sendError(message, "start_debug", "IDE_COMMAND_FAILED", error.message ?: error.javaClass.name)
            }
            return
        }

        val sourcePath = message.filePath ?: message.file
        if (!sourcePath.isNullOrBlank() && message.line != null) {
            startDebugFromSourceLine(message, sourcePath, message.line)
            return
        }

        sendError(message, "start_debug", "INVALID_ARGUMENT", "Pass runConfigName or filePath + line.")
    }

    private fun startDebugFromSourceLine(message: BridgeMessage, sourcePath: String, line: Int) {
        try {
            val target = findConfigurationForSourceLine(sourcePath, line)
            if (target == null) {
                sendError(
                    message,
                    "start_debug",
                    "RUN_CONFIG_NOT_FOUND",
                    "No runnable debug configuration was found at $sourcePath:$line."
                )
                return
            }
            val runManager = RunManager.getInstance(project)
            if (!runManager.hasSettings(target.settings)) {
                runManager.setTemporaryConfiguration(target.settings)
            }
            executeWithOrigin(target.settings, message)
        } catch (error: Throwable) {
            sendError(message, "start_debug", "IDE_COMMAND_FAILED", error.message ?: error.javaClass.name)
        }
    }

    private fun findConfigurationForSourceLine(sourcePath: String, line: Int): SourceDebugTarget? {
        return ApplicationManager.getApplication().runReadAction<SourceDebugTarget?> {
            val virtualFile = LocalFileSystem.getInstance().findFileByPath(sourcePath)
                ?: return@runReadAction null
            val psiFile = PsiManager.getInstance(project).findFile(virtualFile)
                ?: return@runReadAction null
            val document = FileDocumentManager.getInstance().getDocument(virtualFile)
                ?: return@runReadAction null
            val zeroBasedLine = (line - 1).coerceAtLeast(0)
            if (zeroBasedLine >= document.lineCount) return@runReadAction null
            val offset = document.getLineStartOffset(zeroBasedLine)
            var element: PsiElement? = psiFile.findElementAt(offset) ?: psiFile
            while (element != null) {
                val context = ConfigurationContext(element)
                val existing = context.findExisting()
                if (existing != null) {
                    return@runReadAction SourceDebugTarget(existing, element.textRange?.toString() ?: element.javaClass.simpleName)
                }
                val created = context.createConfigurationsFromContext()
                    ?.map { it.configurationSettings }
                    ?.firstOrNull()
                if (created != null) {
                    return@runReadAction SourceDebugTarget(created, element.textRange?.toString() ?: element.javaClass.simpleName)
                }
                element = element.parent
            }
            null
        }
    }

    private fun listRunConfigurations(message: BridgeMessage) {
        val sourcePath = message.filePath ?: message.file
        if (!sourcePath.isNullOrBlank()) {
            listRunPoints(message, sourcePath)
            return
        }
        val configurations = RunManager.getInstance(project).allSettings.map { settings ->
            mapOf(
                "name" to settings.name,
                "description" to settings.type.displayName,
                "supportsDynamicLaunchOverrides" to true
            )
        }
        bridge.send(
            BridgeMessage(
                type = MessageTypes.IdeRunConfigurationsSnapshot,
                requestId = message.requestId,
                sessionId = message.sessionId,
                ideSessionId = message.ideSessionId,
                originRequestId = message.originRequestId,
                pauseEpoch = message.expectedPauseEpoch,
                result = mapOf("configurations" to configurations)
            )
        )
    }

    private fun listRunPoints(message: BridgeMessage, sourcePath: String) {
        val file = LocalFileSystem.getInstance().findFileByPath(sourcePath)
        if (file == null) {
            sendRunConfigurationsError(message, "WORKSPACE_VIOLATION", "File was not found in IDEA local filesystem.")
            return
        }
        val document = FileDocumentManager.getInstance().getDocument(file)
        if (document == null) {
            sendRunConfigurationsError(message, "WORKSPACE_VIOLATION", "File cannot be opened as a document.")
            return
        }
        val runPoints = mutableListOf<Map<String, Any?>>()
        val seenElements = mutableSetOf<String>()
        for (lineIndex in 0 until document.lineCount) {
            val target = findConfigurationForSourceLine(sourcePath, lineIndex + 1) ?: continue
            if (!seenElements.add(target.sourceElement)) continue
            runPoints += mapOf(
                "line" to lineIndex + 1,
                "description" to "Run '${target.settings.name}'\nDebug '${target.settings.name}'",
                "elementText" to target.settings.configuration.name
            )
        }
        bridge.send(
            BridgeMessage(
                type = MessageTypes.IdeRunConfigurationsSnapshot,
                requestId = message.requestId,
                sessionId = message.sessionId,
                ideSessionId = message.ideSessionId,
                originRequestId = message.originRequestId,
                pauseEpoch = message.expectedPauseEpoch,
                result = mapOf(
                    "filePath" to sourcePath,
                    "runPoints" to runPoints
                )
            )
        )
    }

    private fun sendRunConfigurationsError(message: BridgeMessage, code: String, text: String) {
        bridge.send(
            BridgeMessage(
                type = MessageTypes.IdeRunConfigurationsSnapshot,
                requestId = message.requestId,
                sessionId = message.sessionId,
                ideSessionId = message.ideSessionId,
                originRequestId = message.originRequestId,
                pauseEpoch = message.expectedPauseEpoch,
                error = mapOf(
                    "code" to code,
                    "message" to text
                )
            )
        )
    }

    private fun execute(message: BridgeMessage, command: String, action: (com.intellij.xdebugger.XDebugSession) -> Unit) {
        val session = tracker.find(message.ideSessionId)
        if (session == null) {
            sendError(message, command, "IDE_SESSION_NOT_FOUND", "IDE debug session was not found.")
            return
        }
        try {
            tracker.armOrigin(message.ideSessionId, message.originRequestId ?: message.requestId)
            action(session)
            sendResult(message, command, mapOf("ok" to true))
        } catch (error: Throwable) {
            sendError(message, command, "IDE_COMMAND_FAILED", error.message ?: error.javaClass.name)
        }
    }

    private fun runToLine(message: BridgeMessage) {
        val sourcePath = message.filePath ?: message.file
        val line = message.line
        if (sourcePath.isNullOrBlank() || line == null || line < 1) {
            sendError(message, "run_to_line", "INVALID_ARGUMENT", "filePath and line are required.")
            return
        }
        execute(message, "run_to_line") { session ->
            val file = LocalFileSystem.getInstance().findFileByPath(sourcePath)
                ?: throw IllegalArgumentException("File was not found in IDEA local filesystem: $sourcePath")
            val position = XDebuggerUtil.getInstance().createPosition(file, line - 1)
                ?: throw IllegalArgumentException("No executable source position was found at $sourcePath:$line")
            session.runToPosition(position, false)
        }
    }

    private fun setVariable(message: BridgeMessage) {
        val ref = message.ref as? String
        if (ref != null) {
            variableReader.setNativeValue(message.ideSessionId, ref, message.newValue, message.expectedPauseEpoch) { result, error ->
                if (error != null) sendError(message, "set_variable", error, error)
                else sendResult(message, "set_variable", result ?: emptyMap())
            }
            return
        }
        if (message.path.isEmpty() || message.newValue.isNullOrBlank()) {
            sendError(message, "set_variable", "INVALID_ARGUMENT", "path and newValue are required.")
            return
        }
        val target = message.path.joinToString(".")
        val assignment = "$target = ${message.newValue}"
        variableReader.evaluate(message.ideSessionId, assignment) { result, error ->
            if (error != null) {
                sendError(message, "set_variable", "SET_VARIABLE_FAILED", error)
            } else {
                sendResult(
                    message,
                    "set_variable",
                    mapOf(
                        "path" to message.path,
                        "newValue" to message.newValue,
                        "applied" to true,
                        "result" to result
                    )
                )
            }
        }
    }

    private fun evaluate(message: BridgeMessage) {
        val expression = message.expression
            ?: message.payload["expression"] as? String
            ?: message.options["expression"] as? String
            ?: message.result["expression"] as? String
            ?: message.command?.takeIf { it != "evaluate" }
        if (expression.isNullOrBlank()) {
            sendError(message, "evaluate", "INVALID_ARGUMENT", "Expression is required.")
            return
        }
        variableReader.evaluate(message.ideSessionId, expression) { result, error ->
            if (error != null) {
                sendError(message, "evaluate", "EVALUATE_FAILED", error)
            } else {
                sendResult(message, "evaluate", result ?: emptyMap())
            }
        }
    }

    private fun sendResult(message: BridgeMessage, command: String, result: Map<String, Any?>) {
        bridge.send(
            BridgeMessage(
                type = MessageTypes.IdeCommandResult,
                requestId = message.requestId,
                sessionId = message.sessionId,
                ideSessionId = message.ideSessionId,
                originRequestId = message.originRequestId,
                pauseEpoch = message.expectedPauseEpoch,
                command = command,
                result = result
            )
        )
    }

    private fun sendError(message: BridgeMessage, command: String, code: String, text: String) {
        bridge.send(
            BridgeMessage(
                type = MessageTypes.IdeCommandResult,
                requestId = message.requestId,
                sessionId = message.sessionId,
                ideSessionId = message.ideSessionId,
                originRequestId = message.originRequestId,
                pauseEpoch = message.expectedPauseEpoch,
                command = command,
                error = mapOf(
                    "code" to code,
                    "message" to text
                )
            )
        )
    }

    private fun executeWithOrigin(settings: RunnerAndConfigurationSettings, message: BridgeMessage) {
        val environment = ExecutionEnvironmentBuilder
            .create(DefaultDebugExecutor.getDebugExecutorInstance(), settings)
            .build()
        environment.putUserData(BreakPilotExecutionOrigin.key, message.originRequestId ?: message.requestId)
        ProgramRunnerUtil.executeConfiguration(environment, false, true)
    }

    private fun requestStack(message: BridgeMessage) {
        val session = tracker.find(message.ideSessionId)
        val epoch = tracker.pauseEpoch(message.ideSessionId)
        if (session == null || epoch == null) {
            sendError(message, "request_stack", "IDE_SESSION_NOT_FOUND", "IDE debug session was not found.")
            return
        }
        if (message.expectedPauseEpoch != epoch) {
            sendError(message, "request_stack", "STALE_RUNTIME_HANDLE", "Stack request belongs to another paused state.")
            return
        }
        StackReader().read(session, message.threadId, message.offset ?: 0, message.limit ?: 20, epoch) { page ->
            bridge.send(
                BridgeMessage(
                    type = MessageTypes.IdeStackSnapshot,
                    requestId = message.requestId,
                    sessionId = message.sessionId,
                    ideSessionId = message.ideSessionId,
                    originRequestId = message.originRequestId,
                    pauseEpoch = epoch,
                    result = page
                )
            )
        }
    }
}

File: breakpilot-vscode/src/debugger/CommandExecutor.ts
import * as vscode from "vscode";
import { BridgeClient } from "../bridge/BridgeClient";
import { AnyRecord, BridgeMessage, MessageTypes } from "../bridge/MessageProtocol";
import { DebugSessionTracker } from "./DebugSessionTracker";
import { VariableReader } from "./VariableReader";
import { StackReader } from "./StackReader";

export class CommandExecutor {
  constructor(
    private readonly bridge: BridgeClient,
    private readonly tracker: DebugSessionTracker,
    private readonly variableReader: VariableReader
  ) {}

  async handle(message: BridgeMessage) {
    switch (message.type) {
      case MessageTypes.AgentStartDebug:
        await this.startDebug(message);
        break;
      case MessageTypes.AgentRequestStack:
        await this.readStack(message);
        break;
      case MessageTypes.AgentContinue:
        await this.executeDebugCommand(message, "continue", "workbench.action.debug.continue");
        break;
      case MessageTypes.AgentPause:
        await this.executeDebugCommand(message, "pause", "workbench.action.debug.pause");
        break;
      case MessageTypes.AgentStepOver:
        await this.executeDebugCommand(message, "step_over", "workbench.action.debug.stepOver");
        break;
      case MessageTypes.AgentStepInto:
        await this.executeDebugCommand(message, "step_into", "workbench.action.debug.stepInto");
        break;
      case MessageTypes.AgentStepOut:
        await this.executeDebugCommand(message, "step_out", "workbench.action.debug.stepOut");
        break;
      case MessageTypes.AgentRunToLine:
        await this.runToLine(message);
        break;
      case MessageTypes.AgentSetVariable:
        await this.setVariable(message);
        break;
      case MessageTypes.AgentStopDebug:
        await this.stopDebug(message);
        break;
      case MessageTypes.AgentEvaluate:
        await this.evaluate(message);
        break;
      case MessageTypes.AgentListRunConfigurations:
        await this.listRunConfigurations(message);
        break;
    }
  }

  private async executeDebugCommand(message: BridgeMessage, command: string, vscodeCommand: string) {
    const session = this.targetSession(message);
    if (!session) {
      this.sendError(message, command, "IDE_SESSION_NOT_FOUND", "VS Code debug session was not found.");
      return;
    }
    if (vscode.debug.activeDebugSession !== session) {
      this.sendError(
        message,
        command,
        "SESSION_NOT_ACTIVE",
        "VS Code can only continue or step the active debug session from the extension API."
      );
      return;
    }
    try {
      this.tracker.armOrigin(message.ideSessionId, message.originRequestId ?? message.requestId);
      await vscode.commands.executeCommand(vscodeCommand);
      this.sendResult(message, command, { ok: true });
    } catch (error) {
      this.sendError(message, command, "IDE_COMMAND_FAILED", this.errorMessage(error));
    }
  }

  private async stopDebug(message: BridgeMessage) {
    const session = this.targetSession(message);
    if (!session) {
      this.sendError(message, "stop_debug", "IDE_SESSION_NOT_FOUND", "VS Code debug session was not found.");
      return;
    }
    try {
      await vscode.debug.stopDebugging(session);
      this.sendResult(message, "stop_debug", { ok: true });
    } catch (error) {
      this.sendError(message, "stop_debug", "IDE_COMMAND_FAILED", this.errorMessage(error));
    }
  }

  private async evaluate(message: BridgeMessage) {
    const expression = this.expressionFromMessage(message);
    if (!expression) {
      this.sendError(message, "evaluate", "INVALID_ARGUMENT", "Expression is required.");
      return;
    }
    const response = await this.variableReader.evaluate(message.ideSessionId, expression, {
      ...(message.options ?? {}),
      ...(message.payload ?? {}),
      frameId: message.frameId ?? message.options?.frameId ?? message.payload?.frameId,
      threadId: message.threadId ?? message.options?.threadId ?? message.payload?.threadId
    });
    if (response.error) {
      this.sendError(message, "evaluate", "EVALUATE_FAILED", response.error);
      return;
    }
    this.sendResult(message, "evaluate", response.result ?? {});
  }

  private async setVariable(message: BridgeMessage) {
    const response = await this.variableReader.setVariable(message.ideSessionId, message.path, message.newValue, {
      ...(message.options ?? {}),
      ...(message.payload ?? {}),
      frameId: message.frameId ?? message.options?.frameId ?? message.payload?.frameId,
      threadId: message.threadId ?? message.options?.threadId ?? message.payload?.threadId
    }, message.ref);
    if (response.error) {
      const code = response.error === "STALE_RUNTIME_HANDLE" || response.error === "VARIABLE_NOT_MUTABLE"
        ? response.error
        : "SET_VARIABLE_FAILED";
      this.sendError(message, "set_variable", code, response.error);
      return;
    }
    this.sendResult(message, "set_variable", response.result ?? {});
  }

  private async runToLine(message: BridgeMessage) {
    const filePath = typeof message.filePath === "string" ? message.filePath : message.file;
    if (!filePath || !message.line) {
      this.sendError(message, "run_to_line", "INVALID_ARGUMENT", "filePath and line are required.");
      return;
    }
    try {
      const uri = vscode.Uri.file(filePath);
      const document = await vscode.workspace.openTextDocument(uri);
      const editor = await vscode.window.showTextDocument(document, { preview: false });
      const position = new vscode.Position(Math.max(0, message.line - 1), 0);
      editor.selection = new vscode.Selection(position, position);
      await vscode.commands.executeCommand("editor.debug.action.runToCursor");
      this.sendResult(message, "run_to_line", {
        status: "running",
        position: { filePath, line: message.line }
      });
    } catch (error) {
      this.sendError(message, "run_to_line", "RUN_TO_LINE_FAILED", this.errorMessage(error));
    }
  }

  private async listRunConfigurations(message: BridgeMessage) {
    if (message.filePath || message.file) {
      this.bridge.send({
        type: MessageTypes.IdeRunConfigurationsSnapshot,
        requestId: message.requestId,
        sessionId: message.sessionId,
        ideSessionId: message.ideSessionId,
        originRequestId: message.originRequestId,
        pauseEpoch: message.expectedPauseEpoch,
        result: {
          filePath: message.filePath ?? message.file,
          runPoints: []
        }
      });
      return;
    }
    const configurations = vscode.workspace
      .getConfiguration("launch")
      .get<AnyRecord[]>("configurations", [])
      .map((configuration) => ({
        name: String(configuration.name ?? ""),
        description: String(configuration.type ?? "VS Code Debug Configuration"),
        supportsDynamicLaunchOverrides: false
      }))
      .filter((configuration) => configuration.name.length > 0);
    this.bridge.send({
      type: MessageTypes.IdeRunConfigurationsSnapshot,
      requestId: message.requestId,
      sessionId: message.sessionId,
      ideSessionId: message.ideSessionId,
      originRequestId: message.originRequestId,
      pauseEpoch: message.expectedPauseEpoch,
      result: { configurations }
    });
  }

  private async startDebug(message: BridgeMessage) {
    const configurations = vscode.workspace.getConfiguration("launch").get<AnyRecord[]>("configurations", []);
    const runConfigName = typeof message.runConfigName === "string" ? message.runConfigName : undefined;
    const selected = runConfigName
      ? configurations.find((configuration) => configuration.name === runConfigName)
      : configurations[0];
    if (!selected) {
      this.sendError(message, "start_debug", "RUN_CONFIG_NOT_FOUND", "VS Code launch configuration was not found.");
      return;
    }
    const folder = vscode.workspace.workspaceFolders?.[0];
    const configuration: vscode.DebugConfiguration = {
      ...selected,
      type: String(selected.type ?? ""),
      name: String(selected.name ?? runConfigName ?? "BreakPilot"),
      request: String(selected.request ?? "launch"),
      __breakpilotOriginRequestId: message.originRequestId ?? message.requestId
    };
    const started = await vscode.debug.startDebugging(folder, configuration);
    if (!started) {
      this.sendError(message, "start_debug", "IDE_COMMAND_FAILED", "VS Code rejected the debug start request.");
    }
  }

  private async readStack(message: BridgeMessage) {
    const session = this.targetSession(message);
    const pauseEpoch = this.tracker.pauseEpoch(message.ideSessionId);
    if (!session || pauseEpoch === undefined) {
      this.sendError(message, "request_stack", "IDE_SESSION_NOT_FOUND", "VS Code debug session was not found.");
      return;
    }
    if (message.expectedPauseEpoch !== pauseEpoch) {
      this.sendError(message, "request_stack", "STALE_RUNTIME_HANDLE", "Stack request belongs to another paused state.");
      return;
    }
    const threadId = typeof message.threadId === "number" ? message.threadId : this.tracker.sessionInfo(message.ideSessionId)?.threadId;
    if (threadId === undefined) {
      this.sendError(message, "request_stack", "INVALID_ARGUMENT", "A paused thread is required.");
      return;
    }
    try {
      const page = await new StackReader(session).read(threadId, message.offset ?? 0, message.limit ?? 20, pauseEpoch);
      this.bridge.send({
        type: MessageTypes.IdeStackSnapshot,
        requestId: message.requestId,
        sessionId: message.sessionId,
        ideSessionId: message.ideSessionId,
        originRequestId: message.originRequestId,
        pauseEpoch,
        result: page
      });
    } catch (error) {
      this.sendError(message, "request_stack", "STACK_READ_FAILED", this.errorMessage(error));
    }
  }

  private targetSession(message: BridgeMessage): vscode.DebugSession | undefined {
    return this.tracker.find(message.ideSessionId);
  }

  private expressionFromMessage(message: BridgeMessage): string | undefined {
    const candidates = [
      message.expression,
      message.payload?.expression,
      message.options?.expression,
      message.result?.expression,
      message.command && message.command !== "evaluate" ? message.command : undefined
    ];
    return candidates.find((value): value is string => typeof value === "string" && value.trim().length > 0)?.trim();
  }

  private sendResult(message: BridgeMessage, command: string, result: AnyRecord) {
    this.bridge.send({
      type: MessageTypes.IdeCommandResult,
      requestId: message.requestId,
      sessionId: message.sessionId,
      ideSessionId: message.ideSessionId,
      originRequestId: message.originRequestId,
      pauseEpoch: message.expectedPauseEpoch,
      command,
      result
    });
  }

  private sendError(message: BridgeMessage, command: string, code: string, text: string) {
    this.bridge.send({
      type: MessageTypes.IdeCommandResult,
      requestId: message.requestId,
      sessionId: message.sessionId,
      ideSessionId: message.ideSessionId,
      originRequestId: message.originRequestId,
      pauseEpoch: message.expectedPauseEpoch,
      command,
      error: {
        code,
        message: text
      }
    });
  }

  private errorMessage(error: unknown): string {
    return error instanceof Error ? error.message : String(error);
  }
}
